L1 and L2 Regularization




In [1]:
# import libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns


In [2]:
# Suppress Warnings for clean notebook
import warnings
warnings.filterwarnings('ignore')

We are going to use Melbourne House Price Dataset where we'll
predict House Predictions based on various features.

The Dataset Link is
https://www.kaggle.com/anthonypino/melbourne-housing-market

In [3]:
# read dataset
dataset = pd.read_csv('/home/ujjwal/Data_Science/Jupyter_notebook/csv_xlxs_File/Melbourne_housing_FULL.csv')


In [4]:
dataset.head()



Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3        NaN  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [5]:
dataset.nunique()

Suburb             351
Address          34009
Rooms               12
Type                 3
Price             2871
Method               9
SellerG            388
Date                78
Distance           215
Postcode           211
Bedroom2            15
Bathroom            11
Car                 15
Landsize          1684
BuildingArea       740
YearBuilt          160
CouncilArea         33
Lattitude        13402
Longtitude       14524
Regionname           8
Propertycount      342
dtype: int64

In [6]:
dataset.shape


(34857, 21)

In [7]:
# let's use limited columns which makes more sense for serving our purpose
cols_to_use = ['Suburb', 'Rooms', 'Type', 'Method', 'SellerG', 'Regionname', 'Propertycount', 
               'Distance', 'CouncilArea', 'Bedroom2', 'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'Price']
dataset = dataset[cols_to_use]

In [8]:
dataset.head(20)

Suburb  Rooms Type Method  SellerG             Regionname  \
0   Abbotsford      2    h     SS   Jellis  Northern Metropolitan   
1   Abbotsford      2    h      S   Biggin  Northern Metropolitan   
2   Abbotsford      2    h      S   Biggin  Northern Metropolitan   
3   Abbotsford      3    u     VB   Rounds  Northern Metropolitan   
4   Abbotsford      3    h     SP   Biggin  Northern Metropolitan   
5   Abbotsford      3    h     PI   Biggin  Northern Metropolitan   
6   Abbotsford      4    h     VB   Nelson  Northern Metropolitan   
7   Abbotsford      4    h     SN   Nelson  Northern Metropolitan   
8   Abbotsford      2    h      S   Biggin  Northern Metropolitan   
9   Abbotsford      2    h      S  Collins  Northern Metropolitan   
10  Abbotsford      2    h      S   Jellis  Northern Metropolitan   
11  Abbotsford      3    h      S   Nelson  Northern Metropolitan   
12  Abbotsford      2    u     PI   Biggin  Northern Metropolitan   
13  Abbotsford      4    h      W   Biggin  Northern Metropolitan   
14  Abbotsford      2    h      S   Nelson  Northern Metropolitan   
15  Abbotsford      3    h      S   Jellis  Northern Metropolitan   
16  Abbotsford      2    t      S   Jellis  Northern Metropolitan   
17  Abbotsford      1    u      S   Biggin  Northern Metropolitan   
18  Abbotsford      2    h      S   Biggin  Northern Metropolitan   
19  Abbotsford      2    u      S   Dingle  Northern Metropolitan   

    Propertycount  Distance         CouncilArea  Bedroom2  Bathroom  Car  \
0          4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
1          4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
2          4019.0       2.5  Yarra City Council       2.0       1.0  0.0   
3          4019.0       2.5  Yarra City Council       3.0       2.0  1.0   
4          4019.0       2.5  Yarra City Council       3.0       2.0  0.0   
5          4019.0       2.5  Yarra City Council       3.0       2.0  1.0   
6          4019.0       2.5  Yarra City Council       3.0       1.0  2.0   
7          4019.0       2.5  Yarra City Council       3.0       2.0  2.0   
8          4019.0       2.5  Yarra City Council       4.0       1.0  2.0   
9          4019.0       2.5  Yarra City Council       3.0       2.0  1.0   
10         4019.0       2.5  Yarra City Council       2.0       1.0  0.0   
11         4019.0       2.5  Yarra City Council       4.0       2.0  0.0   
12         4019.0       2.5  Yarra City Council       2.0       2.0  1.0   
13         4019.0       2.5  Yarra City Council       6.0       2.0  0.0   
14         4019.0       2.5  Yarra City Council       2.0       1.0  2.0   
15         4019.0       2.5  Yarra City Council       NaN       NaN  NaN   
16         4019.0       2.5  Yarra City Council       NaN       NaN  NaN   
17         4019.0       2.5  Yarra City Council       1.0       1.0  1.0   
18         4019.0       2.5  Yarra City Council       3.0       1.0  2.0   
19         4019.0       2.5  Yarra City Council       NaN       NaN  NaN   

    Landsize  BuildingArea      Price  
0      126.0           NaN        NaN  
1      202.0           NaN  1480000.0  
2      156.0          79.0  1035000.0  
3        0.0           NaN        NaN  
4      134.0         150.0  1465000.0  
5       94.0           NaN   850000.0  
6      120.0         142.0  1600000.0  
7      400.0         220.0        NaN  
8      201.0           NaN        NaN  
9      202.0           NaN        NaN  
10     181.0           NaN   941000.0  
11     245.0         210.0  1876000.0  
12    4292.0          82.0        NaN  
13     230.0         147.0        NaN  
14     256.0         107.0  1636000.0  
15       NaN           NaN  1000000.0  
16       NaN           NaN   745000.0  
17       0.0           NaN   300000.0  
18     220.0          75.0  1097000.0  
19       NaN           NaN   542000.0

In [9]:
dataset.shape


(34857, 15)

Checking for Nan values




In [10]:
dataset.isna().sum()


Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Regionname           3
Propertycount        3
Distance             1
CouncilArea          3
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
Price             7610
dtype: int64

Handling Missing values



In [11]:
# Some feature's missing values can be treated as zero (another class for NA values or absence of that feature)
# like 0 for Propertycount, Bedroom2 will refer to other class of NA values
# like 0 for Car feature will mean that there's no car parking feature with house
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']
dataset[cols_to_fill_zero] = dataset[cols_to_fill_zero].fillna(0)

# other continuous features can be imputed with mean for faster results since our focus is on Reducing overfitting
# using Lasso and Ridge Regression
dataset['Landsize'] = dataset['Landsize'].fillna(dataset.Landsize.mean())
dataset['BuildingArea'] = dataset['BuildingArea'].fillna(dataset.BuildingArea.mean())


Drop NA values of Price, since it's our predictive variable we won't impute it



In [12]:
dataset.dropna(inplace=True)


In [13]:
dataset.shape


(27244, 15)

Let's one hot encode the categorical features

In [14]:
dataset = pd.get_dummies(dataset, drop_first=True)
#here we remove all the text value , see here for   'CouncilArea'

In [15]:
dataset.head()


Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car  Landsize  \
1      2         4019.0       2.5       2.0       1.0  1.0     202.0   
2      2         4019.0       2.5       2.0       1.0  0.0     156.0   
4      3         4019.0       2.5       3.0       2.0  0.0     134.0   
5      3         4019.0       2.5       3.0       2.0  1.0      94.0   
6      4         4019.0       2.5       3.0       1.0  2.0     120.0   

   BuildingArea      Price  Suburb_Aberfeldie  ...  \
1      160.2564  1480000.0                  0  ...   
2       79.0000  1035000.0                  0  ...   
4      150.0000  1465000.0                  0  ...   
5      160.2564   850000.0                  0  ...   
6      142.0000  1600000.0                  0  ...   

   CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
1                                    0                                  0   
2                                    0                                  0   
4                                    0                                  0   
5                                    0                                  0   
6                                    0                                  0   

   CouncilArea_Nillumbik Shire Council  CouncilArea_Port Phillip City Council  \
1                                    0                                      0   
2                                    0                                      0   
4                                    0                                      0   
5                                    0                                      0   
6                                    0                                      0   

   CouncilArea_Stonnington City Council  CouncilArea_Whitehorse City Council  \
1                                     0                                    0   
2                                     0                                    0   
4                                     0                                    0   
5                                     0                                    0   
6                                     0                                    0   

   CouncilArea_Whittlesea City Council  CouncilArea_Wyndham City Council  \
1                                    0                                 0   
2                                    0                                 0   
4                                    0                                 0   
5                                    0                                 0   
6                                    0                                 0   

   CouncilArea_Yarra City Council  CouncilArea_Yarra Ranges Shire Council  
1                               1                                       0  
2                               1                                       0  
4                               1                                       0  
5                               1                                       0  
6                               1                                       0  

[5 rows x 745 columns]

Let's bifurcate our dataset into train and test dataset

In [16]:
X = dataset.drop('Price', axis=1)
y = dataset['Price']


In [17]:
from sklearn.model_selection import train_test_split
train_X, test_X, train_y, test_y = train_test_split(X, y, test_size=0.3, random_state=2)

Let's train our Linear Regression Model on training dataset and check the accuracy on test set

In [18]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression().fit(train_X, train_y)


In [19]:
reg.score(test_X, test_y)


0.13853683161589492

In [20]:
reg.score(train_X, train_y)


0.6827792395792723

Here training score is 68% but test score is 13.85% which is very low

Normal Regression is clearly overfitting the data, let's try other models


Using Lasso (L1 Regularized) Regression Model

In [23]:
from sklearn import linear_model
lasso_reg = linear_model.Lasso(alpha=50, max_iter=100, tol=0.1)
lasso_reg.fit(train_X, train_y)


Lasso(alpha=50, max_iter=100, tol=0.1)

In [24]:
lasso_reg.score(test_X, test_y)


0.6636111369404488

In [26]:
lasso_reg.score(train_X, train_y)


0.6766985624766824

Using Ridge (L2 Regularized) Regression Model

In [27]:
from sklearn.linear_model import Ridge
ridge_reg= Ridge(alpha=50, max_iter=100, tol=0.1)
ridge_reg.fit(train_X, train_y)


Ridge(alpha=50, max_iter=100, tol=0.1)

In [28]:

ridge_reg.score(test_X, test_y)


0.6670848945194958

In [29]:
ridge_reg.score(train_X, train_y)


0.6622376739684328

In [ ]:
We see that Lasso and Ridge Regularizations prove to be beneficial when our Simple
Linear Regression Model overfits. These results may not be that contrast but significant 
in most cases.Also that L1 & L2 Regularizations are used in Neural Networks too